pkg/bpf/bpffs.go

```diff
@@ -104,100 +118,275 @@ func OpenAfterMount(m *Map) error {
-		log.WithField(logfields.Path, mapRoot).Warningf("BPF root is not a BPF filesystem (%#x != %#x)",
-			uint32(fsdata.Type), magic)
+
+	if uint32(fsdata.Type) == magic {
+		return true, true, nil
+	}
+	return true, false, nil
+}
+
+// checkOrMountCustomLocation tries to check or mount the BPF filesystem in the
+// given path.
+func checkOrMountCustomLocation(bpfRoot string) error {
+	SetMapRoot(bpfRoot)
+
+	// Check whether the custom location has a BPFFS mount.
+	mounted, bpffsInstance, err := isBPFFS()
+	if err != nil {
+		return err
+	}
+
+	// If the custom location has no mount, let's mount BPFFS there.
+	if !mounted {
+		SetMapRoot(bpfRoot)
+		if err := mountFS(); err != nil {
+			return err
+		}
+	}
+
+	// If the custom location already has a mount with some other filesystem than
+	// BPFFS, return the error.
+	if !bpffsInstance {
+		return fmt.Errorf("mount in the custom directory %s has a different filesystem than BPFFS", bpfRoot)
+	}
+
+	return nil
+}
+
+// checkOrMountCustomLocation tries to check or mount the BPF filesystem in
+// standard locations, which are:
+// - /sys/fs/bpf
+// - /run/cilium/bpffs
+// There is a procedure of determining which directory is going to be used:
+// 1. Checking whether BPFFS filesystem is mounted in /sys/fs/bpf.
+// 2. If there is no mount, then mount BPFFS in /sys/fs/bpf and finish there.
+// 3. If there is a BPFFS mount, finish there.
+// 4. If there is a mount, but with the other filesystem, then it means that most
+//    probably Cilium is running inside container which has mounted /sys/fs/bpf
+//    from host, but host doesn't have proper BPFFS mount, so that mount is just
+//    the empty directory. In that case, mount BPFFS under /run/cilium/bpffs.
+func checkOrMountDefaultLocations() error {
+	var err error
+
+	// Check whether /sys/fs/bpf has a BPFFS mount.
+	mounted, bpffsInstance, err := isBPFFS()
+	if err != nil {
+		return err
+	}
+
+	// If /sys/fs/bpf is not mounted at all, we should mount
+	// BPFFS there.
+	if !mounted {
+		err = mountFS()
+	} else if !bpffsInstance {
+		// If /sys/fs/bpf has a mount but with some other filesystem
+		// than BPFFS, it means that Cilium is running inside container and
+		// /sys/fs/bpf is not mounted on host.
+		// We should mount BPFFS in /run/cilium/bpffs automatically.
+		// This will allow operation of Cilium but will result in unmounting of
+		// the filesystem when the pod is restarted. This in turn will cause resources
+		// such as the connection tracking table of the BPF programs to be released
+		// which will cause all connections into local containers to be dropped.
+		// User is going to be warned about that.
+		log.Warnf("BPF filesystem is going to be mounted automatically in %s. However, it probably "+
+			"means that Cilium is running inside container and BPFFS is not mounted on "+
+			"the host. We recommend to do that: "+
+			"https://cilium.readthedocs.io/en/latest/kubernetes/install/?highlight=bpffs#mounting-the-bpf-fs-optional",
```

Comment:
This probably needs a short URL similar the one with have for the kvstore error cc @tgraf